python/gs/ops/sddmm.py

```diff
@@ -0,0 +1,180 @@
+from itertools import product
+from .sparse import *
+import sys
+
+__all__ = ['gsddmm', 'copy_u', 'copy_v', 'copy_e']
+
+
+def reshape_lhs_rhs(lhs_data, rhs_data):
+    r""" Expand dims so that there will be no broadcasting issues with different
+    number of dimensions. For example, given two shapes (N, 3, 1), (E, 5, 3, 4)
+    that are valid broadcastable shapes, change them to (N, 1, 3, 1) and
+    (E, 5, 3, 4)
+
+    Parameters
+    ----------
+    lhs_data : tensor or None
+        The left operand, could be None if it's not required by op.
+    rhs_data : tensor or None
+        The right operand, could be None if it's not required by op.
+    """
+    lhs_shape = lhs_data.shape
+    rhs_shape = rhs_data.shape
+    if len(lhs_shape) != len(rhs_shape):
+        max_ndims = max(len(lhs_shape), len(rhs_shape))
+        lhs_pad_ndims = max_ndims - len(lhs_shape)
+        rhs_pad_ndims = max_ndims - len(rhs_shape)
+        new_lhs_shape = (lhs_shape[0],) + (1,) * lhs_pad_ndims + lhs_shape[1:]
+        new_rhs_shape = (rhs_shape[0],) + (1,) * rhs_pad_ndims + rhs_shape[1:]
+        lhs_data = lhs_data.view(new_lhs_shape)
+        rhs_data = rhs_data.view(new_rhs_shape)
+    return lhs_data, rhs_data
+
+
+def gsddmm(g, op, lhs_data, rhs_data, lhs_target='u', rhs_target='v'):
```

Comment:
Again, I feel the `lhs_target` and `rhs_target` are not necessary.

Reply:
This is necessary, since lhs_target can be `v` or `e` not just `u`, which determines where the index of data we fetch from, `col_id`, `row_id` or `edge_id`.